[Collapse] Migrate to emotion

vinyldarkscratch opened the pull request:
This PR migrates the `Collapse` component for v5 as a part of #24405.

**Notice:** this is my very first pull request to this repository.  If I have made any mistakes, please do not hesitate to let me know and I will fix it right away!  Feel free to give any and all critique -- I'm a big girl, I can handle it.  😛 

- [X] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

_This contribution, and all other contributions I make to Material UI, are performed under [Gooborg Studios, LLC](https://www.gooborg.com) on behalf of [Fox and Geese, LLC](https://github.com/foxandgeese).  Both companies agree to surrender all contributions under MUI's MIT license._

vinyldarkscratch commented:
There's two tests that I'm not really sure how to fix at this moment, nor do I feel that disabling the tests is the right call.  The failing tests are the following:

```
  1) <StepContent />
       renders children inside an Collapse component:
     TypeError: Cannot read property 'root' of null
      at Context.<anonymous> (packages/material-ui/src/StepContent/StepContent.test.js:46:66)
      at processImmediate (node:internal/timers:463:21)

  2) <StepContent />
       prop: transitionDuration
         should use default Collapse component:
     TypeError: Cannot read property 'root' of null
      at Context.<anonymous> (packages/material-ui/src/StepContent/StepContent.test.js:66:68)
      at processImmediate (node:internal/timers:463:21)
```

Would love some help on this one!

mnajdova commented on packages/material-ui/src/Collapse/collapseClasses.js:
```diff
@@ -0,0 +1,14 @@
+
+const collapseClasses = generateUtilityClasses('MuiCollapse', [
+  'root',
+  'horizontal',
```
```suggestion
  'horizontal',
  'vertical',
```

mnajdova commented on packages/material-ui/src/Collapse/Collapse.js:
```diff
@@ -238,24 +293,27 @@ const Collapse = React.forwardRef(function Collapse(props, ref) {
-            <div
+            <CollapseWrapperInner
+              styleProps={styleProps}
               className={clsx(classes.wrapperInner, {
```
Let's add the horizontal class on this slot in the `useUtilityClasses` and have here only `className={classes.wrapperInner}` Same for the `wrapper`

mnajdova commented on packages/material-ui/src/Collapse/Collapse.js:
```diff
@@ -2,71 +2,117 @@ import * as React from 'react';
+
+  const slots = {
+    root: ['root', `${orientation}`],
+    wrapper: ['wrapper'],
```
```suggestion
    wrapper: ['wrapper', `${orientation}`],
```

mnajdova commented on packages/material-ui/src/Collapse/Collapse.js:
```diff
@@ -2,71 +2,117 @@ import * as React from 'react';
+  const slots = {
+    root: ['root', `${orientation}`],
+    wrapper: ['wrapper'],
+    wrapperInner: ['wrapperInner'],
```
```suggestion
    wrapperInner: ['wrapperInner', `${orientation}`],
```

mnajdova reviewed (commented):
Really great job so far, especially that this is your first PR on the repo 🚀 This component is a bit tricky 😄 I've tried to leave comments for everything I found. Let's address these comments and test the component, and then we can move to do fixes for the tests.

Again, great job!

mnajdova commented on packages/material-ui/src/Collapse/Collapse.js:
```diff
@@ -81,6 +127,14 @@ const Collapse = React.forwardRef(function Collapse(props, ref) {
   } = props;
+
+  const styleProps = {
+    ...other,
```
```suggestion
    ...props,
```

I overlooked this, sorry :)

vinyldarkscratch commented:
Thank you very much for the quick review, @mnajdova!  I've committed all of the suggested changes (they were very educational, thank you), and will soon be addressing the remaining review comments. 😄

mnajdova commented:
> Thank you very much for the quick review, @mnajdova! I've committed all of the suggested changes (they were very educational, thank you), and will soon be addressing the remaining review comments. 😄

Sure thing! If you have some doubt, take the `Button` component as a reference, or take a look at the [migration guide](https://gist.github.com/mnajdova/b0562785b7cf1b9d9aeae382d578594b). I'll check the PR again after all comments are resolved

mnajdova commented on packages/material-ui/src/Collapse/Collapse.js:
```diff
@@ -81,6 +130,16 @@ const Collapse = React.forwardRef(function Collapse(props, ref) {
+    ...props,
+    orientation,
+    collapsedSize: collapsedSizeProp,
+    in: inProp,
```
```suggestion
```

Seems like this will be already included in the props and it does not have a default value, so we can safely drop it. The thinking here is that, the `stylesProps` should include all props + their defaults, that is why we are spreading the props and then overriting with the default values

vinyldarkscratch commented:
Alright, all review comments have been addressed, thank you very much!  All that remains now are just those issues with the tests for `StepContent`.

mnajdova commented:
> Alright, all review comments have been addressed, thank you very much! All that remains now are just those issues with the tests for StepContent.

We need to change how the `collapseClasses` are being used.  I believe this diff should fix the test issues:

```diff
diff --git a/packages/material-ui/src/StepContent/StepContent.test.js b/packages/material-ui/src/StepContent/StepContent.test.js
index 9266055ee4..dabdfe9a0d 100644
--- a/packages/material-ui/src/StepContent/StepContent.test.js
+++ b/packages/material-ui/src/StepContent/StepContent.test.js
@@ -1,20 +1,18 @@
 import * as React from 'react';
 import { expect } from 'chai';
 import { getClasses, createClientRender, createMount, describeConformance } from 'test/utils';
-import Collapse from '../Collapse';
+import Collapse, { collapseClasses } from '../Collapse';
 import Stepper from '../Stepper';
 import Step from '../Step';
 import StepContent from './StepContent';

 describe('<StepContent />', () => {
   let classes;
-  let collapseClasses;
   const mount = createMount({ strict: true });
   const render = createClientRender();

   before(() => {
     classes = getClasses(<StepContent />);
-    collapseClasses = getClasses(<Collapse />);
   });

   describeConformance(<StepContent />, () => ({

```

vinyldarkscratch commented:
Oh durr, that makes total sense...  🤦‍♀️   Patch applied, and the tests are now passing, thank you!

mnajdova commented:
I believe this line is not correct https://github.com/mui-org/material-ui/pull/24501/files#diff-0b07fe6763d1a29535d5239ffd35dd227da26c69e664e0d01c0619414b0e86a0R81 it should probably be

```
      const collapse = container.querySelector(`.${collapseClasses.root}`);
```

@oliviertassinari what do you think?

mnajdova reviewed (approved):
Great first PR on Material-UI 🚀
